### ISSUE TITLE
CBMC dumps core with invariant violation: offset bits are byte aligned

### ISSUE THREAD

**Issue by danielsn** [diffblue/cbmc#5344]
CBMC version: `5.12 (cbmc-5.12-225-g6e1ded2c7)`
Operating system: `Linux ip-172-31-81-124 5.3.0-1017-aws #18~18.04.1-Ubuntu SMP Wed Apr 8 15:12:16 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`
Exact command line resulting in the issue: `make report`
What behaviour did you expect: CBMC to verify the code
What happened instead: CBMC crashed:
```
Generated 9611 VCC(s), 8741 remaining after simplification
Passing problem to propositional reduction
converting SSA
--- begin invariant violation report ---
Invariant check failed
File: ../src/solvers/lowering/byte_operators.cpp:1252 function: lower_byte_extract
Condition: subtype_bits.has_value() && *subtype_bits == 8
Reason: offset bits are byte aligned
Backtrace:
cbmc(print_backtrace(std::ostream&)+0x4d) [0x55940e4fb752]
cbmc(get_backtrace[abi:cxx11]()+0x45) [0x55940e4fb959]
cbmc(std::enable_if<std::is_base_of<invariant_failedt, invariant_failedt>::value, void>::type invariant_violated_structured<invariant_failedt, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x4e) [0x55940ddd852b]
cbmc(+0x7a5cd7) [0x55940ddd7cd7]
cbmc(lower_byte_extract(byte_extract_exprt const&, namespacet const&)+0x1cbe) [0x55940e329ed2]
cbmc(boolbvt::convert_byte_extract(byte_extract_exprt const&)+0xcc) [0x55940e2c2f36]
cbmc(boolbvt::convert_bitvector(exprt const&)+0xeda) [0x55940e2b4d42]
cbmc(bv_pointerst::convert_bitvector(exprt const&)+0xe52) [0x55940e2ffdf6]
cbmc(boolbvt::convert_bv(exprt const&, nonstd::optional_lite::optional<unsigned long>)+0x100) [0x55940e2b3832]
cbmc(boolbvt::convert_equality(equal_exprt const&)+0x2d5) [0x55940e2cc951]
cbmc(boolbvt::convert_rest(exprt const&)+0x1e8) [0x55940e2b6574]
cbmc(bv_pointerst::convert_rest(exprt const&)+0x89f) [0x55940e2fb001]
cbmc(prop_conv_solvert::convert_bool(exprt const&)+0x159c) [0x55940e3412b2]
cbmc(prop_conv_solvert::convert(exprt const&)+0x194) [0x55940e33fc2a]
cbmc(prop_conv_solvert::convert_bool(exprt const&)+0x4f0) [0x55940e340206]
cbmc(prop_conv_solvert::convert(exprt const&)+0x194) [0x55940e33fc2a]
cbmc(prop_conv_solvert::set_equality_to_true(equal_exprt const&)+0xad) [0x55940e3417eb]
cbmc(prop_conv_solvert::add_constraints_to_prop(exprt const&, bool)+0x5b1) [0x55940e341e3b]
cbmc(prop_conv_solvert::set_to(exprt const&, bool)+0x52) [0x55940e3428a2]
cbmc(boolbvt::set_to(exprt const&, bool)+0x1bc) [0x55940e2b8582]
cbmc(decision_proceduret::set_to_true(exprt const&)+0x2c) [0x55940e2a364a]
cbmc(symex_target_equationt::convert_assignments(decision_proceduret&)+0x12a) [0x55940e04c34c]
cbmc(symex_target_equationt::convert_without_assertions(decision_proceduret&)+0x78) [0x55940e04c0f6]
cbmc(symex_target_equationt::convert(decision_proceduret&)+0x23) [0x55940e04c1c3]
cbmc(convert_symex_target_equation(symex_target_equationt&, decision_proceduret&, message_handlert&)+0x91) [0x55940de085cf]
cbmc(prepare_property_decider(std::unordered_map<dstringt, property_infot, std::hash<dstringt>, std::equal_to<dstringt>, std::allocator<std::pair<dstringt const, property_infot> > >&, symex_target_equationt&, goto_symex_property_decidert&, ui_message_handlert&)+0x108) [0x55940de09987]
cbmc(multi_path_symex_checkert::prepare_property_decider(std::unordered_map<dstringt, property_infot, std::hash<dstringt>, std::equal_to<dstringt>, std::allocator<std::pair<dstringt const, property_infot> > >&)+0x4f) [0x55940de1d53f]
cbmc(multi_path_symex_checkert::operator()(std::unordered_map<dstringt, property_infot, std::hash<dstringt>, std::equal_to<dstringt>, std::allocator<std::pair<dstringt const, property_infot> > >&)+0x1d3) [0x55940de1d3f9]
cbmc(all_properties_verifier_with_trace_storaget<multi_path_symex_checkert>::operator()()+0x55) [0x55940ddf5855]
cbmc(cbmc_parse_optionst::doit()+0x17e0) [0x55940dde1cf8]
cbmc(parse_options_baset::main()+0x117) [0x55940e52085b]
cbmc(main+0x55) [0x55940ddd799f]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f9a6bd8ab97]
cbmc(_start+0x2a) [0x55940ddd786a]


--- end invariant violation report ---
cbmc  --flush --bounds-check --conversion-check --div-by-zero-check --float-overflow-check --nan-check --pointer-check --pointer-overflow-check --pointer-primitive-check --signed-overflow-check --undefined-shift-check --unsigned-overflow-check --unwind 1 --unwinding-assertions --unwindset aws_cryptosdk_edk_list_clear.0:2,aws_cryptosdk_edk_list_elements_are_bounded.0:2,aws_cryptosdk_edk_list_elements_are_valid.0:2,aws_cryptosdk_edk_list_is_bounded.0:2,aws_cryptosdk_edk_list_is_valid.0:2,aws_cryptosdk_keyring_trace_clear.0:2,aws_cryptosdk_keyring_trace_is_valid.0:2,ensure_cryptosdk_edk_list_has_allocated_list_elements.0:2,ensure_cryptosdk_edk_list_has_allocated_members.0:2,ensure_trace_has_allocated_records.0:2 --object-bits 8 --show-properties --xml-ui aws_cryptosdk_enc_materials_destroy_harness.goto 2>&1 > property.xml
```

Code located at: https://github.com/danielsn/aws-encryption-sdk-c/tree/ESDK-enc-materials-destroy-invarient-violation
Proof: aws_cryptosdk_enc_materials_destroy
command: `make report`

**Comment by thomasspriggs** [comment]
I have been unable to reproduce this issue, due to the following build error -
```
$ make report
mkdir -p link-farm
echo "Setting up dependencies"
Setting up dependencies
make: *** No rule to make target '/home/tspriggs/Development/aws-encryption-sdk-c/.cbmc-batch/jobs/aws_cryptosdk_enc_materials_destroy/link-farm/c-common-helper-src/make_common_data_structures.goto', needed by 'aws_cryptosdk_enc_materials_destroy_harness0.goto'. Stop.
```
Do you have a simpler (standalone) example which reproduces this invariant violation?

**Comment by hannes-steffenhagen-diffblue** [comment]
Tracked internally as ADA-571

**Comment by danielsn** [comment]
I can try to get that example working again

**Comment by hannes-steffenhagen-diffblue** [comment]
@danielsn Any updates on this? Without a reproducer we would have to close this.

**Comment by tautschnig** [comment]
I just confirmed that this had the same root cause the problem reported in #5718, and was therefore also fixed by #5750. I was able to reproduce the original problem on the branch posted by @danielsn, and able to confirm the fix. Closing.
